packages/language/src/syntax-tree/ast-iterator.ts

```diff
@@ -931,3 +931,53 @@ export function forEachNode(
       break;
   }
 }
+
+export function traverseAllNodes(node: SyntaxNode, traverse: (n: SyntaxNode) => void) {
+  traverse(node);
+  forEachNode(node, (child) => {
+    traverseAllNodes(child as SyntaxNode, traverse);
+  });
+}
```

Comment:
Suggestion: We should be able to return from this traversal early in case it is no longer required, like is the case in `containsNodeOfKind`. For example via something like this:
```suggestion
export enum TraversalState {
  Continue,
  Stop,
}

export function traverseAllNodes(node: SyntaxNode, traverse: (n: SyntaxNode) => TraversalState | undefined | void) {
  const state = traverse(node);
  if (state !== TraversalState.Stop) {
    forEachNode(node, (child) => {
      traverseAllNodes(child as SyntaxNode, traverse);
    });
  }
}
```